### PR TITLE
updates for 0.6

### PR DESCRIPTION
**Pull request by bramtayl**
Changes for 0.6, plus some very limited cleanups.

### CODE REVIEW

**Comment by bramtayl** [comment]
Well, I got 0.5 and 0.6 working; maybe I'll take another stab tomorrow on 0.7

File: .travis.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ language: julia
 julia:
   - 0.4
   - 0.5
+  - 0.6
```
Comment (nalimilan) [inline .travis.yml]: Can you also add the badge to README.md?

File: src/DataFramesMeta.jl [hunk 2]
```diff
@@ -182,13 +172,15 @@ where(d::AbstractDataFrame, arg) = d[arg, :]
 where(d::AbstractDataFrame, f::Function) = d[f(d), :]
 where(g::GroupedDataFrame, f::Function) = g[Bool[f(x) for x in g]]
 
-collect_ands(x::Expr) = x
-collect_ands(x::Expr, y::Expr) = :($x & $y)
-collect_ands(x::Expr, y...) = :($x & $(collect_ands(y...)))
+and(x, y) =
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: I'd say just drop support for versions before 0.6, anyway we already have a working version for these releases. @ararslan Does that sound OK to you?

File: src/DataFramesMeta.jl [hunk 3]
```diff
@@ -43,27 +46,39 @@ replace_syms!(e::Expr, membernames) =
 
 protect_replace_syms!(e, membernames) = e
 protect_replace_syms!(e::Expr, membernames) =
-    e.head == :quote ? e : replace_syms!(e, membernames)
+    if e.head == :quote
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: The old version as a one-liner was more readable IMHO. If you change that you should use the long `function` syntax, but that doesn't sound like a good idea.

File: src/DataFramesMeta.jl [hunk 4]
```diff
@@ -28,6 +29,8 @@ onearg(e, f) = e.head == :call && length(e.args) == 2 && e.args[1] == f
 mapexpr(f, e) = Expr(e.head, map(f, e.args)...)
 
 replace_syms!(x, membernames) = x
+replace_syms!(q::QuoteNode, membernames) =
+    replace_syms!( Meta.quot(q.value), membernames)
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Remove space.

File: src/DataFramesMeta.jl [hunk 5]
```diff
@@ -1,6 +1,7 @@
 module DataFramesMeta
 
 using DataFrames
+import DataFrames
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Why is this needed?

File: src/DataFramesMeta.jl [hunk 6]
```diff
@@ -43,27 +46,39 @@ replace_syms!(e::Expr, membernames) =
...
+            :( $d[$key] )
         end
-        $funname($(funargs...))
+        Expr(:call, anon(values(membernames), body), funargs...)
     end
 end
 
+function with_anonymous(body)
+    d = gensym()
+    anon( d, with_helper(d, body) )
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Remove spaces too, that's not the style used in the file AFAICT.

File: src/DataFramesMeta.jl [hunk 7]
```diff
@@ -182,13 +172,15 @@ where(d::AbstractDataFrame, arg) = d[arg, :]
...
+    if VERSION < v"0.6.0-"
+        :($x & $y)
+    else
+        :($x .& $y)
+    end
 
 function where_helper(d, args...)
-    with_args = :($DataFramesMeta.@with(_DF, $(collect_ands(args...))))
-    :( $DataFramesMeta.where($d, _DF -> $with_args) )
+    :( $where($d, $(with_anonymous( reduce(and, args)))))
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Same here (spaces are only used inside `:(`, not after any parenthesis).

File: src/DataFramesMeta.jl [hunk 8]
```diff
@@ -475,17 +448,27 @@ function Base.select(d::Union{AbstractDataFrame, Associative}; kwargs...)
...
 end
 
-
-function select_helper(x, args...)
-    select_args = :(select(_DF; $(expandargs(args)...)))
-    quote
-        let _DF = $x
-            DataFramesMeta.@with(_DF, $select_args)
-        end
+replace_equals_with_kw(e) =
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Same style remark here and below: either use ternary operator, or `function`.

File: src/byrow.jl [hunk 9]
```diff
@@ -51,31 +51,27 @@ function byrow_helper(df, body)
     (e_body, e_newcols) = byrow_find_newcols(body, Any[])
     e_delimiters = Expr(:(=), :row, :( 1:_N ))
     e_forloop = Expr(:for, e_delimiters, byrow_replace(e_body))
-    return quote
-        _N = length($df[1])
-        _DF = @transform($df, $(e_newcols...))
-        @with _DF $e_forloop
-        _DF
-    end
+    :(_N = length($df[1]); _DF = @transform($df, $(e_newcols...)); @with _DF $e_forloop; _DF)
```
Comment (nalimilan) [inline src/byrow.jl]: Is this really more readable than the previous form?

File: src/byrow.jl [hunk 10]
```diff
@@ -51,31 +51,27 @@ function byrow_helper(df, body)
...
 Act on a DataFrame row-by-row.
 
 Includes support for control flow and `begin end` blocks. Since the
-"environment" induced by `@byrow! df` is implicitly a single row of `df`, 
+"environment" induced by `@byrow! df` is implicitly a single row of `df`,
 use regular operators and comparisons instead of their elementwise counterparts
 as in `@with`. Note that the scope within `@byrow!` is a hard scope.
 
 `byrow!` also supports special syntax for allocating new columns. The syntax
-`@newcol x::Array{Int}` allocates a new column `:x` with an `Array` container
```
Comment (nalimilan) [inline src/byrow.jl]: Are you sure this is right? IIRC the `::` syntax was intended.

File: src/compositedataframe.jl [hunk 11]
```diff
@@ -43,14 +43,18 @@ A constructor of an `AbstractCompositeDataFrame` that mimics the `DataFrame`
 constructor.  This returns a composite type (not immutable) that is an
 `AbstractCompositeDataFrame`.
 
-This uses `eval` to create a new type within the current module.
+This uses `eval` to create a new type within the module specified by the
+inmodule keyword.
```
Comment (nalimilan) [inline src/compositedataframe.jl]: "keyword *argument"*. Use backticks around `inmodule`. Should also mention `inmodule = DataFramesMeta` before `kwargs...` in signatures.

File: src/compositedataframe.jl [hunk 12]
```diff
@@ -43,14 +43,18 @@ A constructor of an `AbstractCompositeDataFrame` that mimics the `DataFrame`
...
 
 ### Arguments
 
 * `columns` : contains the contents of the columns
 * `cnames` : the names of the columns
 * `typename` : the optional name of the type created
 * `kwargs` : the key gives the column names, and the value is the column contents
+* `inmodule = DataFramesMeta` : a keyword argument to specify what module you
+want to define the type in. Consider passing `current_module()` or
```
Comment (nalimilan) [inline src/compositedataframe.jl]: Align this line and the next to `` `inmodule` `` AFAIK. If you drop support for Julia < 0.6, no need to mention old Julia versions.

File: test/compositedataframes.jl [hunk 13]
```diff
@@ -55,4 +57,26 @@ for x in eachrow(df)
...
+
+@test names(hcat(df, df2)) == [:A, :B, :C, :D]
+@test names(hcat(df2, df)) == [:C, :D, :A, :B]
+@test names(hcat(df, df3)) == [:A, :B, :C, :D]
+
+@test CompositeDataFrame(DataFrame(df)) == df
+@test names(CompositeDataFrame(DataFrame(df), [:C, :D])) == [:C, :D]
+@test row() == nothing
+@test df[1, :A] == 1
+@test names( df[[1, 2]] ) == [:A, :B]
```
Comment (nalimilan) [inline test/compositedataframes.jl]: No spaces (here and below). `[1, 2]` could be `1:2`.

File: test/dataframes.jl [hunk 14]
```diff
@@ -36,8 +37,8 @@ idx2 = :B
 @test  @where(df, :A .> x)          == df[df[:A] .> x,:]
 @test  @where(df, :B .> x)          == df[df[:B] .> x,:]
 @test  @where(df, :A .> :B)         == df[df[:A] .> df[:B],:]
-@test  @where(df, :A .> 1, :B .> 1) == df[(df[:A] .> 1) & (df[:B] .> 1),:]
-@test  @where(df, :A .> 1, :A .< 4, :B .> 1) == df[(df[:A] .> 1) & (df[:B] .> 1),:]
+@test  @where(df, :A .> 1, :B .> 1) == df[map(&, df[:A] .> 1, df[:B] .> 1),:]
```
Comment (nalimilan) [inline test/dataframes.jl]: If you drop support for Julia < 0.6, use `.&`.

File: test/dataframes.jl [hunk 15]
```diff
@@ -53,13 +54,15 @@ y = 0
...
-    @newcol colY::DataArray{Int}
+    @newcol colX = Array{Float64}(_N)
+    @newcol colY = DataArray(Int, _N)
     :colX = :B == 2 ? pi * :A : :B
     if :A > 1
         :colY = :A * :B
     end
 end
+
+import MacroTools
```
Comment (nalimilan) [inline test/dataframes.jl]: ?

File: test/runtests.jl [hunk 16]
```diff
@@ -1,4 +1,4 @@
-fatalerrors = false
+fatalerrors = true
```
Comment (nalimilan) [inline test/runtests.jl]: Probably better keep `false` here, as it allows running all tests to see all errors, and only fail at the end.

File: .travis.yml (hunk 1 above)
Comment (nalimilan) [inline .travis.yml]: Actually the badge is already there.

File: src/DataFramesMeta.jl (hunk 3 above)
Comment (bramtayl) [inline src/DataFramesMeta.jl]: I like if else better than ternary because ternary gives artificially inflated coverage.

File: src/DataFramesMeta.jl (hunk 2 above)
Comment (bramtayl) [inline src/DataFramesMeta.jl]: Oh no! I spent all that time getting it working on 0.4 :(

File: src/byrow.jl (hunk 10 above)
Comment (bramtayl) [inline src/byrow.jl]: In fact, :: (or any two arg expr) should work as well. However, I did have to change it so that the user needs to manually use _N. This was because the deprecation of the Array(T, size) syntax.

File: test/compositedataframes.jl (hunk 13 above)
Comment (bramtayl) [inline test/compositedataframes.jl]: I was trying hit coverage of the `Base.getindex{T <: DataFrames.ColumnIndex}(cdf::AbstractCompositeDataFrame, col_inds::AbstractVector{T})` constructor

File: src/byrow.jl (hunk 9 above)
Comment (bramtayl) [inline src/byrow.jl]: This is on purpose. The multi-line quote blocks will introduce extraneous line number information into user code.

File: src/compositedataframe.jl (hunk 12 above)
Comment (bramtayl) [inline src/compositedataframe.jl]: `current_module()` is only deprecated as of 0.7.

File: src/byrow.jl (hunk 10 above)
Comment (nalimilan) [inline src/byrow.jl]: I don't understand. What is `_N`? Wouldn't `::Vector{Int}` work since we're dealing with columns?

File: src/compositedataframe.jl (hunk 12 above)
Comment (nalimilan) [inline src/compositedataframe.jl]: OK. Would make sense to mention the versions then, since in a few months the old version won't be relevant at all.

File: src/byrow.jl (hunk 9 above)
Comment (nalimilan) [inline src/byrow.jl]: Just curious, what kind of information?

File: src/DataFramesMeta.jl (hunk 3 above)
Comment (nalimilan) [inline src/DataFramesMeta.jl]: I wouldn't say that's a sufficient reason. If think it's been discussed in Base already.

Anyway, when you use `if` blocks, please also use the long `function` form (there are several cases like this), and keep the short syntax for one-liners.

File: src/compositedataframe.jl [hunk 17]
```diff
@@ -33,24 +33,28 @@ row() = nothing
 
 """
 ```julia
-CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol})
-CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}, typename::Symbol)
-CompositeDataFrame(; kwargs...)
-CompositeDataFrame(typename::Symbol; kwargs...)
+CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}; inmodule = DataFramesMeta)
+CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}, typename::Symbol; inmodule = DataFramesMeta)
```
Comment (nalimilan) [inline src/compositedataframe.jl]: This line needs to be wrapped at 92 characters.

File: test/grouping.jl [hunk 18]
```diff
@@ -14,9 +14,11 @@ g = groupby(d, :x)
 @test  DataFrame(@where(g, length(:x) > 5))[:n][1:3] == @data [5, 6, 7]
 
 @test  DataFrame(orderby(g, x -> mean(x[:n]))) == DataFrame(@orderby(g, mean(:n)))
-#@test  DataFrames.based_on(@orderby(g, mean(:n)), x -> x[1,:x])[:x1] == [3,1,2]
```
Comment (nalimilan) [inline test/grouping.jl]: Why remove this? Is it tested elsewhere? Else, better keep it as it might reflect a feature which is supposed to work but doesn't.

Comment (bramtayl) [inline test/grouping.jl]: I added in a `@based_on` test which should cover this.

File: src/byrow.jl (hunk 10 above)
Comment (nalimilan) [inline src/byrow.jl]: Scratch that, `_N` is not the dimension of the array as I first thought. Though it still don't understand the change.

Comment (bramtayl) [inline src/byrow.jl]: Can you initialize an empty column in a dataframe (or subdataframe) of a certain type without using the length of the dataframe?

Comment (bramtayl) [inline src/byrow.jl]: The `@newcol` feature might be more trouble than it's worth; it could simply be accomplished by a transform. Giving transform (or actually, any with function) access to _N would be useful, similar to how n() is used in dplyr.

Comment (nalimilan) [inline src/byrow.jl]: IIRC the point was to support a syntax similar to SAS's `data` step, which is quite practical when doing many operations. Having to specify `_N` manually goes against the goal of having a concise syntax, and that doesn't make a lot of sense since "new column" necessarily means "a vector with a length equal to the number of rows of the data frame". Do you have a use case in mind where `_N` would be useful in `@byrow!`? Note that this command is not as flexible as the other ones.

Comment (bramtayl) [inline src/byrow.jl]: Well the reason I changed it in the first place is the deprecation of `Array(T, size)`. The macro in the previous version was taking `Container{Eltype}` and turning it into `Container(Eltype, _N)`.  This used to work for both Arrays and DataArrays. Now I could have changed it to make instead `Container{Eltype}(_N)`, which would have been the correct syntax for Arrays but not for  DataArrays. Tldr. the macro was making unwarranted assumptions about constructors, which was why it was able to leave out _N.

File: src/byrow.jl (hunk 9 above)
Comment (bramtayl) [inline src/byrow.jl]: Well, user code after a macroexpand would contain line numbers from the byrow.jl file. Actually, I guess that's probably a good thing for backtraces?, but maybe due to a bug, these lines would show up as uncovered. Right now coverage for this pull request is up to like 97% in part because of changes I made like this.

File: src/byrow.jl (hunk 10 above)
Comment (nalimilan) [inline src/byrow.jl]: OK, but that's a bug in DataArrays, so let's use `Container{Eltype}(_N)`, avoid testing this with `DataArray`, and this the problem in that package.

File: src/byrow.jl (hunk 9 above)
Comment (nalimilan) [inline src/byrow.jl]: Yes, I'd rather get good backtraces than nominally high coverage. It's not a big deal if these lines appear as uncovered while they are covered. That's a bug which should be fixed elsehwere, let's not make the codebase worse because of this.

Comment (bramtayl) [inline src/byrow.jl]: P.S. this macro (as with all macros here) are completely escaped such that both _N and _DF will end up in user code. I was originally thinking this was a bug, because a user could get unexpected results when referencing these variables, but in fact, an aspiring user could use this strategically. Maybe it deserves documentation?

Comment (nalimilan) [inline src/byrow.jl]: Yeah, ideally they should be documented (unless we escape them, but keeping them doesn't sound too dangerous and could be useful).

Comment (bramtayl) [inline src/byrow.jl]: The best of both worlds would be to replace all references to line numbers in generated code with the line numbers in user code. (Hopefully) if there's an error to find it should be in user code and not in package code. I have a NumberedLines package which can do this. Dunno if we want to add another dependency.

Comment (nalimilan) [inline src/byrow.jl]: That would clearly be ideal, but currently the future of this package is not completely clear AFAIK, so I'm not sure it's worth spending too much time on new features before knowing where we're heading to. Though if the dependency is not too large it could be worth it.

Comment (bramtayl) [inline src/byrow.jl]: This would also require __source__ which is only available in 0.7

**Comment by bramtayl** [comment]
Got to pass on 0.7. Looks like `@>` from Lazy is broken.

File: src/DataFramesMeta.jl [hunk 19]
```diff
@@ -207,18 +219,72 @@ Multiple `i` expressions are "and-ed" together.
...
+│ Row │ x │ y │
+├─────┼───┼───┤
+│ 1   │ 2 │ 1 │
+│ 2   │ 3 │ 2 │
+
+julia> @where(df, :x .> x, :y .== 3)
+0×2 DataFrames.DataFrame
+
+julia> d = DataFrame(
+            n = 1:20,
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: That's an unusual way of wrapping lines. I'd rather keep `n` on the first line, but split `x` over multiple lines. Same remark in the next docstring.

File: src/DataFramesMeta.jl (hunk 2 above)
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Please switching this to the long `function` form.

File: src/DataFramesMeta.jl [hunk 20]
```diff
@@ -253,19 +319,14 @@ orderbyconstructor(d::AbstractDataFrame) = (x...) -> DataFrame(Any[x...])
...
 function orderby_helper(d, args...)
-    construct_args = :(DataFramesMeta.orderbyconstructor(_D)($(args...)))
-    with_args = :(DataFramesMeta.@with(_DF, $construct_args))
-    quote
-        let _D = $d
-            DataFramesMeta.orderby(_D, _DF -> $with_args)
-        end
-    end
+    _D = gensym()
+    :(let $_D = $d
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Why not keep `quote... end`? Looks cleaner.

File: src/DataFramesMeta.jl [hunk 21]
```diff
@@ -274,14 +335,39 @@ Sort by criteria. Normally used to sort groups in GroupedDataFrames.
...
+```jldoctest
+julia> using DataFrames, DataFramesMeta
+
+julia> d = DataFrame(
+            n = 1:20,
+            x = [3, 3, 3, 3, 1, 1, 1, 2, 1, 1, 2, 1, 1, 2, 2, 2, 3, 1, 1, 2]);
+
+julia> g = groupby(d, :x);
+
+julia> @macroexpand @orderby(g, mean(:n))
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: What does `@macroexpand` do here?

File: src/DataFramesMeta.jl [hunk 22]
```diff
@@ -320,22 +406,14 @@ function transform(g::GroupedDataFrame; kwargs...)
...
-    kw.args[2] = :( _DF -> DataFramesMeta.@with(_DF, $(kw.args[2]) ) )
-    kw
-end
-
 function transform_helper(x, args...)
-    # convert each kw arg value to: _DF -> @with(_DF, arg)
-    newargs = [args...]
-    map!(convert_kw!, newargs)
-    :( transform($x, $(newargs...)) )
+    :( $transform($x, $(map(args) do kw
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: `quote... end` could make sense here too.

File: src/DataFramesMeta.jl [hunk 23]
```diff
@@ -411,14 +536,12 @@ end
 ##############################################################################
 
 function by_helper(x, what, args...)
-    with_args = :(DataFramesMeta.@with(_DF, DataFrames.DataFrame($(args...))))
-    :( DataFrames.by($x, $what, _DF -> $with_args) )
+    :($by($x, $what,
+        $(with_anonymous(:($DataFrame($(map(replace_equals_with_kw, args)...)))))))
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Align with `$x`.

File: src/DataFramesMeta.jl [hunk 24]
```diff
@@ -475,22 +629,33 @@ function Base.select(d::Union{AbstractDataFrame, Associative}; kwargs...)
...
+    if e.head == :(=)
+        Expr(:kw, e.args[1], e.args[2])
+    else
+        e
     end
 end
 
+expandargs(x) = x
+expandargs(q::QuoteNode) = Expr(:kw, q.value, q)
+expandargs(e::Expr) =
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: Use long `function` form here too.

File: src/DataFramesMeta.jl [hunk 25]
```diff
@@ -475,22 +629,33 @@ function Base.select(d::Union{AbstractDataFrame, Associative}; kwargs...)
...
+    if e.head == :quote
+        Expr(:kw, e.args[1], e)
+    else
+        replace_equals_with_kw(e)
+    end
 
+function select_helper(x, args...)
+    DF = gensym()
+    select_args = with_helper(DF, :($select($DF, $(map(expandargs, args)...))))
+    :(let $DF = $x
```
Comment (nalimilan) [inline src/DataFramesMeta.jl]: `quote... end`.

File: src/byrow.jl [hunk 26]
```diff
@@ -28,18 +28,23 @@ byrow_replace(x) = x
 
 function byrow_find_newcols(e::Expr, newcol_decl)
     if e.head == :macrocall && e.args[1] == Symbol("@newcol")
-        ea = e.args[2]
+        ea =
```
Comment (nalimilan) [inline src/byrow.jl]: It's more common to write this as:
```julia
if ...
    ea = ...
else
   ea ...
end
```

File: src/byrow.jl [hunk 27]
```diff
@@ -48,34 +53,34 @@ end
...
 use regular operators and comparisons instead of their elementwise counterparts
 as in `@with`. Note that the scope within `@byrow!` is a hard scope.
 
 `byrow!` also supports special syntax for allocating new columns. The syntax
 `@newcol x::Array{Int}` allocates a new column `:x` with an `Array` container
 with eltype `Int`. Note that the returned `AbstractDataFrame` includes these new
 columns, but the original `d` is not affected. This feature makes it easier to
-use `byrow!` for data transformations.
+use `byrow!` for data transformations. `_N` is introduced to represent the
+length of the dataframe, `_D` represents the dataframe including added columns,
```
Comment (nalimilan) [inline src/byrow.jl]: "data frame". "current row" means index of the current row?

File: src/byrow.jl [hunk 28]
```diff
@@ -89,16 +94,33 @@ The modified `AbstractDataFrame`.
...
-df2 = @byrow! df begin
-    @newcol colX::Array{Float64}
-    :colX = :B == 2 ? pi * :A : :B
-end
+julia> using DataFrames, DataFramesMeta
+
+julia> df = DataFrame(A = 1:3, B = [2, 1, 2]);
+
+julia> let x = 0
+            @byrow!(df, if :A + :B == 3; x += 1 end)  #  This doesn't work without the let
```
Comment (nalimilan) [inline src/byrow.jl]: Would be clearer with `begin... end` block on multiple lines. Same below.

File: src/byrow.jl [hunk 29]
```diff
@@ -89,16 +94,33 @@ The modified `AbstractDataFrame`.
...
+julia> @byrow! df if :A > :B; :A = 0 end
+3×2 DataFrames.DataFrame
+│ Row │ A │ B │
+├─────┼───┼───┤
+│ 1   │ 1 │ 2 │
+│ 2   │ 0 │ 1 │
+│ 3   │ 0 │ 2 │
+
+julia> df2 = @byrow! df begin
+            @newcol colX::Array{Float64}
```
Comment (nalimilan) [inline src/byrow.jl]: Four-space indentation.

File: src/compositedataframe.jl [hunk 30]
```diff
@@ -23,76 +23,89 @@ of the field in `cdf`.
...
 row() = nothing
 
 """
-```julia
-CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol})
-CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}, typename::Symbol)
-CompositeDataFrame(; kwargs...)
-CompositeDataFrame(typename::Symbol; kwargs...)
-```
+    CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}; inmodule = DataFramesMeta)
```
Comment (nalimilan) [inline src/compositedataframe.jl]: Wrap lines at 92 chars.

File: src/linqmacro.jl [hunk 31]
```diff
@@ -27,22 +26,40 @@ The following embedded function calls are equivalent to their macro version:
...
-    by(:b, meanX = mean(:x), meanY = mean(:y)) |>
-    orderby(:meanX) |>
-    select(:meanX, :meanY, var = :b)
+```jldoctest
+julia> using DataFrames, DataFramesMeta
+
+julia> n = 100;
+
+julia> df = DataFrame(a = rand(1:3, n),
+            b = ["a","b","c","d"][rand(1:4, n)],
```
Comment (nalimilan) [inline src/linqmacro.jl]: Align with `a`.

File: src/linqmacro.jl [hunk 32]
```diff
@@ -27,22 +26,40 @@ The following embedded function calls are equivalent to their macro version:
...
+julia> @linq select(orderby(x1, :b, -:meanX), var = :b, :meanX, :meanY)
+3×3 DataFrames.DataFrame
+│ Row │ var │ meanX    │ meanY   │
+├─────┼─────┼──────────┼─────────┤
+│ 1   │ "a" │ 0.665682 │ 6.65682 │
+│ 2   │ "b" │ 0.617848 │ 6.17848 │
+│ 3   │ "d" │ 0.568289 │ 5.68289 │
+
+julia> @linq df |>
+            transform(y = 10 * :x) |>
```
Comment (nalimilan) [inline src/linqmacro.jl]: Four-space indent.

File: .travis.yml [hunk 33]
```diff
@@ -1,7 +1,7 @@
 language: julia
 julia:
-  - 0.4
```
Comment (nalimilan) [inline .travis.yml]: Also remove the README.md badge.

File: src/compositedataframe.jl [hunk 34]
```diff
@@ -38,7 +38,8 @@ row() = nothing
 
 """
     CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}; inmodule = DataFramesMeta)
-    CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}, typename::Symbol; inmodule = DataFramesMeta)
+    CompositeDataFrame(columns::Vector{Any}, cnames::Vector{Symbol}, typename::Symbol;
+        inmodule = DataFramesMeta)
```
Comment (nalimilan) [inline src/compositedataframe.jl]: Should be aligned with `column`.

**Comment by nalimilan** [comment]
Hmm, and you should remove Julia 0.4. from appveyor.yml, which will make the CI turn green.

File: src/DataFramesMeta.jl [hunk 35]
```diff
@@ -182,19 +194,20 @@ where(d::AbstractDataFrame, arg) = d[arg, :]
...
-collect_ands(x::Expr) = x
-collect_ands(x::Expr, y::Expr) = :($x & $y)
-collect_ands(x::Expr, y...) = :($x & $(collect_ands(y...)))
+function and(x, y)
+    if VERSION < v"0.6.0-"
+        :($x & $y)
+    else
+        :($x .& $y)
+    end
+end
```
Comment (ararslan) [inline src/DataFramesMeta.jl]: It would be better to move the conditional outside of the function definition, i.e.
```julia
if VERSION < v"0.6.0-"
    and(x, y) = :($x & $y)
else
    and(x, y) = :($x .& $y)
end
```
Otherwise the version check is performed every time the function is called.

File: src/byrow.jl [hunk 36]
```diff
@@ -28,18 +28,22 @@ byrow_replace(x) = x
 
 function byrow_find_newcols(e::Expr, newcol_decl)
     if e.head == :macrocall && e.args[1] == Symbol("@newcol")
-        ea = e.args[2]
+        if VERSION < v"0.7-"
```
Comment (ararslan) [inline src/byrow.jl]: It would be good to be more exact with the version cutoff used. What changed in Base such that the number of arguments in the expression increased? If you can find the PR that introduced that change, you can use contrib/commit-name.sh in the Julia repo to get the exact version that corresponds to that change.

Comment (ararslan) [inline src/byrow.jl]: Also since this is in a function (but in this case the conditional is less easily separable, the condition should have an `@static` to evaluate the branch once at parse time.

Comment (nalimilan) [inline src/byrow.jl]: Probably https://github.com/JuliaLang/julia/commit/3c3ced4153cdc2c75319b41155c38ad1f2545fc2? That's `0.7.0-DEV.481`.

Comment (bramtayl) [inline src/byrow.jl]: Actually I think it was this: https://github.com/JuliaLang/julia/pull/21746

Comment (bramtayl) [inline src/byrow.jl]: Try `:(@newcol Array{Int}).args` in 0.7

Comment (nalimilan) [inline src/byrow.jl]: That would be 0.7.0-DEV.357 then.

**Comment by nalimilan** [comment]
Can you also remove Julia 0.4 from appveyor.yml?

File: appveyor.yml [hunk 37]
```diff
@@ -1,9 +1,9 @@
 environment:
   matrix:
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.4/julia-0.4-latest-win32.exe"
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.4/julia-0.4-latest-win64.exe"
   - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.5/julia-0.5-latest-win32.exe"
   - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.5/julia-0.5-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.6/julia-0.5-latest-win32.exe"
```
Comment (nalimilan) [inline appveyor.yml]: Typo `0.5` in the filename... :-/

**Comment by nalimilan** [comment]
Thanks! It's really cool to have DataFramesMeta working on Julia 0.6.

**Comment by bramtayl** [comment]
Yeah for sure. I think there's a number of improvements possible going forward. For example, it seems easy enough to widen the scope to include datatables as well. Also, I wonder if anything could be done about type stability/other optimizations for the grouped operations. I've been consider adding DataFramesMeta as a (second) backend to LazyQuery. For some things, like grouped operations, DataFramesMeta seems much more intuitive than Query (which currently doesn't have a good alternative to dplyr summarize).

File: src/compositedataframe.jl [hunk 38]
```diff
@@ -9,9 +9,9 @@ export AbstractCompositeDataFrame, AbstractCompositeDataFrameRow,
 An abstract type that is an `AbstractDataFrame`. Each type that inherits from
 this is expected to be a type-stable data frame.
 """
-abstract AbstractCompositeDataFrame <: AbstractDataFrame
+@compat abstract type AbstractCompositeDataFrame <: AbstractDataFrame end
```
Comment (tkelman) [inline src/compositedataframe.jl]: this requires Compat 0.17
